Makefile

```diff
@@ -70,59 +70,44 @@ coverage: ${ENV_DIR}/bin/coverage cluster_mgr_setup.py queue_mgr_setup.py config
+.PHONY: build-wheelhouse
+build-wheelhouse: ${ENV_DIR}/.cluster-mgr-build-wheelhouse ${ENV_DIR}/.queue-mgr-build-wheelhouse ${ENV_DIR}/.config-mgr-build-wheelhouse
 
-${ENV_DIR}/.queue-mgr-build-eggs: queue_mgr_setup.py shared_setup.py common/setup.py $(shell find src/metaswitch -type f -not -name "*.pyc") $(shell find common/metaswitch -type f -not -name "*.pyc") src/metaswitch/clearwater/queue_manager/alarm_constants.py
-        # Generate .egg files
-	${ENV_DIR}/bin/python queue_mgr_setup.py build -b build_queuemgr bdist_egg -d queue_mgr_eggs
-	${ENV_DIR}/bin/python shared_setup.py build -b build_shared bdist_egg -d queue_mgr_eggs
-	cd common && EGG_DIR=../queue_mgr_eggs make build_common_egg
+define python_component
+# 1 Component Name - dash separated
 
-        # Download the egg files they depend upon
-	${ENV_DIR}/bin/easy_install -zmaxd queue_mgr_eggs/ queue_mgr_eggs/clearwater_queue_manager-1.0-py2.7.egg
-	${ENV_DIR}/bin/easy_install -zmaxd queue_mgr_eggs/ queue_mgr_eggs/clearwater_etcd_shared-1.0-py2.7.egg
-	${ENV_DIR}/bin/easy_install -zmaxd queue_mgr_eggs/ queue_mgr_eggs/metaswitchcommon-0.1-py2.7-linux-x86_64.egg
+build-wheelhouse: ${ENV_DIR}/.$1-build-wheelhouse
 
-	touch $@
+${ENV_DIR}/.$1-build-wheelhouse: $$(subst -,_,$1)_setup.py shared_setup.py common/setup.py $(shell find src/metaswitch -type f -not -name "*.pyc") $(shell find common/metaswitch -type f -not -name "*.pyc") src/metaswitch/clearwater/queue_manager/alarm_constants.py
 
-${ENV_DIR}/.config-mgr-build-eggs: config_mgr_setup.py shared_setup.py common/setup.py $(shell find src/metaswitch -type f -not -name "*.pyc") $(shell find common/metaswitch -type f -not -name "*.pyc") src/metaswitch/clearwater/config_manager/alarm_constants.py
-	# Generate .egg files
-	${ENV_DIR}/bin/python config_mgr_setup.py build -b build_configmgr bdist_egg -d config_mgr_eggs
-	${ENV_DIR}/bin/python shared_setup.py build -b build_shared bdist_egg -d config_mgr_eggs
-	cd common && EGG_DIR=../config_mgr_eggs make build_common_egg
+	rm -f $$@
 
-	# Download the egg files they depend upon
-	${ENV_DIR}/bin/easy_install -zmaxd config_mgr_eggs/ config_mgr_eggs/clearwater_config_manager-1.0-py2.7.egg
-	${ENV_DIR}/bin/easy_install -zmaxd config_mgr_eggs/ config_mgr_eggs/clearwater_etcd_shared-1.0-py2.7.egg
-	${ENV_DIR}/bin/easy_install -zmaxd config_mgr_eggs/ config_mgr_eggs/metaswitchcommon-0.1-py2.7-linux-x86_64.egg
+	# Generate wheels
+	${PYTHON} $$(subst -,_,$1)_setup.py build -b build_queuemgr bdist_wheel -d $$(subst -,_,$1)_wheelhouse
+	${PYTHON} shared_setup.py build -b build_shared bdist_wheel -d $$(subst -,_,$1)_wheelhouse
+	cd common && WHEELHOUSE=../$$(subst -,_,$1)_wheelhouse make build_common_wheel
 
-	touch $@
+	# Download the required dependencies
+	${PIP} wheel -w $$(subst -,_,$1)_wheelhouse -r $$(subst -,_,$1)-requirements.txt -r shared-requirements.txt -r common/requirements.txt --find-links $$(subst -,_,$1)_wheelhouse
```

Comment:
This seems way over-complicated... `pip wheel -w <wheelhouse> <folder>` does all of the above.  You can pin versions in `setup.py` if that was your worry?

Might require python_common/common making wheels themselves, which doesn't seem like a bad thing.

Reply:
python-common will now build a wheel itself - (or rather will when https://github.com/Metaswitch/python-common/pull/107 is merged).

Let's discuss - but in short, no, that doesn't work. `setup.py` pins don't work.

If module A has a pin of C to a specific version, and also depends on version B, which also requires an unpinned version of C, then if `easy_install` processes the B dependency first, then it will install the latest version of C, and then when it later processes the A pin of C, it will install the pinned version.

This means that the directory ends up with both versions in, which then fails at install time.

The order of dependency processing is deterministic, but horrible. We've worked round this in various ways - e.g. https://github.com/Metaswitch/crest/pull/353 and https://github.com/Metaswitch/crest/pull/355 - but the correct fix appears to be using a requirements file which we can specify everything implicity and do the right thing.

Reply:
🐼